### DISCUSSION TITLE
Sending multiple readings from Device Service

### DISCUSSION THREAD

**Discussion by priyadarshi0601** [edgexfoundry/community#44] [Q&A]
Hi Guys.
I had a couple of queries:

1. Is there a way to configure device services like device-mqtt so that it would aggregate edgex readings per device-resource(across a certain interval) and send them to redis or core-data at that certain interval?
2. Can we configure device-mqtt so that it aggregates edgex readings and sends one event consisting of multiple readings of different device-resources at the same timestamp.

Are these features present in Jakarta version?

**Comment by cloudxxx8** [comment]
In the Device Profile, you can define a Device Command which aggregate all the Device Resources you want, and execute the Device Command by API or AutoEvent.
https://docs.edgexfoundry.org/2.3/walk-through/Ch-WalkthroughDeviceProfile/#understanding-device-commands

**Reply by priyadarshi0601** [reply]
Hi.
When I create a device and device profile with the following toml and yml files then I noticed there is a values command also created which only supports get so I assume that it will return all the device-resources in one shot.If so what will be the channel name on redis to which the device-mqtt will send data?

For ex right now I have 2 device resources like temperature and humidity and the topic names are like this:
- edgex.events.device.my-custom-device-profile.my-custom-device.humidity
- edgex.events.device.my-custom-device-profile.my-custom-device.temperature

So what will be the topic name on redis when a get command is performed on values?

Also the way you mentioned that by using device-commands we can aggregate device-resources but wont that be on the device level?Means I want the device-service not the device to aggregate the readings

**Reply by lenny-goodell** [reply]
The last level in the published topic is the `source` which is the `resource name` or `device command name`. So if you use a device command to aggregate multiple resources, the topic will be:

- edgex.events.device.my-custom-device-profile.my-custom-device.my-device-command

**Reply by priyadarshi0601** [reply]
But then the corresponding device should support that aggregate command right?

**Reply by cloudxxx8** [reply]
It depends on the Device Service implementation.  The Device Service can use the aggregate command to retrieve readings, or use loop to retrieve readings one by one and send out in one event.

**Reply by priyadarshi0601** [reply]
If possible can you give an example?On how to write the aggregate command in my.custom.device.profile.yml?
I was following this link https://docs.edgexfoundry.org/2.1/examples/Ch-ExamplesAddingMQTTDevice/

**Comment by lenny-goodell** [comment]
Here is example of profiles device commands that aggregate multiple resources.
https://github.com/edgexfoundry/device-virtual-go/blob/main/cmd/res/profiles/device.virtual.uint.yaml#L138
https://github.com/edgexfoundry/device-mqtt-go/blob/main/cmd/res/profiles/mqtt.test.device.profile.yml#L81

**Reply by priyadarshi0601** [reply]
Thank you

**Reply by priyadarshi0601** [reply]
Just one last question that for all this the device must support the get and set commands right?If it does not do that then it will not work.

**Reply by cloudxxx8** [reply]
the `readWrite` field indicate whether it supports Get or Set commands.  `R` supports get command, and `W` supports set command.  `RW` supports both.
